compiler/util/parser_types.py

```diff
@@ -12,72 +12,139 @@
 
-"""Various types shared through multiple passes of parsing.
+"""Types related to the LR(1) parser.
 
-This module contains types used as interfaces between parts of the Emboss front
-end.  These types do not really "belong" to either the producers or consumers,
-and in a few cases placing them in one or the other creates unnecessary
-dependencies, so they are defined here.
+This module contains types used by the LR(1) parser, which are also used in
+other parts of the compiler: 
+
+    SourcePosition: a position (zero-width) within a source file.
+    SourceLocation: a span within a source file.
+    Production: a grammar production.
+    Token: a token; lr1.Parser operates on lists of tokens.
 """
 
 import collections
-from compiler.util import ir_data
-
-
-def _make_position(line, column):
-    """Makes an ir_data.Position from line, column ints."""
-    if not isinstance(line, int):
-        raise ValueError("Bad line {!r}".format(line))
-    elif not isinstance(column, int):
-        raise ValueError("Bad column {!r}".format(column))
-    return ir_data.Position(line=line, column=column)
 
+PositionTuple = collections.namedtuple(
+    "PositionTuple", ["line", "column"], defaults=[0, 0]
+)
 
-def _parse_position(text):
-    """Parses an ir_data.Position from "line:column" (e.g., "1:2")."""
-    line, column = text.split(":")
-    return _make_position(int(line), int(column))
 
+class SourcePosition(PositionTuple):
+    """A zero-width position within a source file."""
 
-def format_position(position):
-    """formats an ir_data.Position to "line:column" form."""
-    return "{}:{}".format(position.line, position.column)
+    def __new__(cls, /, line=0, column=0):
+        assert isinstance(line, int), f"line {line} is not int"
+        assert isinstance(column, int), f"column {column} is not int"
+        assert line >= 0, f"line {line} is negative"
+        assert column >= 0, f"column {column} is negative"
+        return PositionTuple.__new__(cls, line, column)
 
+    def __str__(self):
+        return f"{self.line}:{self.column}"
 
-def make_location(start, end, is_synthetic=False):
-    """Makes an ir_data.Location from (line, column) tuples or ir_data.Positions."""
-    if isinstance(start, tuple):
-        start = _make_position(*start)
-    if isinstance(end, tuple):
-        end = _make_position(*end)
-    if not isinstance(start, ir_data.Position):
-        raise ValueError("Bad start {!r}".format(start))
-    elif not isinstance(end, ir_data.Position):
-        raise ValueError("Bad end {!r}".format(end))
-    elif start.line > end.line or (
-        start.line == end.line and start.column > end.column
+    @staticmethod
+    def from_str(value):
+        try:
+            l, c = value.split(":")
+            return SourcePosition(line=int(l.strip()), column=int(c.strip()))
+        except Exception as e:
+            raise ValueError(f"{repr(value)} is not a valid SourcePosition.")
+
+    def __bool__(self):
```

Comment:
I'm not a 100% sure adding a bool operator is a good idea or not, see my notes above. A `is_default()` might be more clear about what our intentions are.

Reply:
We cleared this up, this is fine.